records/README.md

```diff
@@ -0,0 +1,220 @@
+
+IPRS records are expressed as [merkledag](../merkledag) objects. This means that the records are linked authenticated data structures, and can be natively replicated over IPFS itself and other merkledag distribution systems.
+
+The objects:
+
+- A `Record` object expresses a value, a validity scheme, and validity data.
+- A `Signature` object could be used to sign and authenticate a record.
+- An `Encryption` object could be used to encrypt a record.
+
+```go
+Record Node {
+  Scheme   Link // link to a validity scheme
+  Value    Link // link to an object representing the value.
+  Validity Data // data needed to satisfy the validity scheme
+}
+```
+
+To achieve good performance, record storage and transfer should bundle all the necessary objects and transmit them together. While "the record object" is only one of the dag objects, "the full record" means a bundle of all objects needed to fully represent, verify, and use the record. (This recommendation does not necessarily include data that records _describe_, for example an ipfs provider record (which signals to consumers that certain data is available) would not include the data itself as part of "the full record").
+
+### The Interface
+
+The IPRS interface is below. It has a few types and functions. We use the Go language to express it, but this is language agnostic.
+
+```go
+// Record is the base type. user can define other types that
+// extend Record.
+type Record struct {
+  Scheme    Link // link to the validity scheme
+  Signature Link // link to a cryptographic signature over the rest of record
+  Value     Data // an opaque value
+}
+
+// Validator is a function that returns whether a record is valid.
+// Users provide their own Validator implementations.
+type Validator func(r *Record) (bool, error)
+
+// Marshal/Unmarshal specifies a way to code the record
+type Marshal(r *Record) ([]byte, error)
+type Unmarshal(r *Record, []byte) (error)
+```
+
+### Interface Example
+
+For example, Alice and Bob earlier could use the following interface:
+
+```protobuf
+type Record struct {
+  Scheme    Link // link to the validity scheme
+  Signature Link // link to a cryptographic signature over the record
+  Expires   Data // datetime at which record expires
+  Value     Data // an opaque value
+}
+
+
+func Validator(r *Record) (bool, error) {
+  authorKey := recordSigningKey(r)
+
+  // always check the signature first
+  sigok := authorKey.Verify(r.Signature, signablePart(r))
+  if !sigok {
+    return false, errors.New("invalid signature. forged record?")
+  }
+
+  // check the expiration.
+  if r.Expires < time.Now() {
+    return false, errors.New("record expired.")
+  }
+
+  return true, nil
+}
+
+func Marshal(r *Record) ([]byte, error) {
+  return recordProtobuf.Marshal(r)
+}
+
+func Unmarshal(r *Record, d []byte) (error) {
+  return recordProtobuf.Unmarshal(r, d)
+}
+```
+
+## Example Record Types
+
+For ease of use, IPRS implementations should include a set of common record types:
```

Comment:
On Fri, May 29, 2015 at 12:08:36PM -0700, W. Trevor King wrote:
> Fri, May 29, 2015 at 12:00:31PM -0700, Juan Batiz-Benet:
> > some records (like dht provider records) are distributed in the
> > ipfs-dht under the key for the content they give a record for,
> > instead of the author's key.
> 
> Oh, right.  I guess they'll need a separate section on how they
> avoid collisions/forgery and aggregate provider records for a given
> resource.

So I'm not sure what the current syntax here is, but I'd expect the
provider-list record would be keyed with something like:

  <multihash-of-provided-object>/providers

the payload itself wouldn't be a signature, but would Merkle-link to
signatures where the link names were the provider IDs.  Each linked
signature would have a signed payload with links:

  Provider: <multihash-of-provider>
  Object: <multihash-of-provided-object>

So the Merkle chain of the record would be:

  record
    <provider-id>: <signature-id>
  ↓               
  signature
    Key: <provider-id>
    Signee: <providing-claim-id>
    …
  ↓
  <providing-claim-id>
    Provider: <provider-id>
    Object: <multihash-of-provided-object>

Nodes that had reason to trust each other (e.g. to not forge providing
claims or to properly currate a provider-list) wouldn't have to fetch
and verify the signed data.  Nodes that had no reason to trust each
other (e.g. fetching a provider list from an untrusted node or
receiving a new providing-claim from an untrusted node) should aquire
and check the signature before using the providing-claim for lookup or
adding it to a provider-list.  Since most nodes won't trust each
other, I expect the signature, providing-claim, public-key, etc.,
packets would be passed around in one optimistic-transmit block [1],
and probably be optimistically hosted on the same nodes that host the
provider-list itself for that purpose.

[1]: https://github.com/ipfs/specs/pull/7#discussion_r31356057